README.md
# TInsul

![TInsul](https://github.com/drericstrong/tinsul/blob/master/images/tinsul_small.png)

[![PyPI version](https://badge.fury.io/py/tinsul.svg)](https://badge.fury.io/py/tinsul)
[![Documentation Status](https://readthedocs.org/projects/tinsul/badge/?version=latest)](http://tinsul.readthedocs.io/en/latest/?badge=latest)

TInsul (Transformer INSULation simulator) is a Python module to simulate condition-monitoring data for the paper insulation of large-scale transformers. It might be used to test new prognostics algorithms, as a teaching aid for those new to the field of condition-monitoring, or to compare the results from simulated data to real-world applications.

TInsul requires the following primary inputs:

* Average temperatures per month
* Starting month (optional)
* Overload ratio (optional)

and will provide these output values (per week):

* CO generation
* CO2 generation
* Furan
* Furfural
* Water Content
* Failure time (inferred from the size of the array)

*Several simplifying assumptions were made when developing TInsul, and empirical data from an academic paper was used to estimate the rate of dissolved gas accumulation per week. While the simulated data from TInsul will be relatively realistic, its real-world applicability should be treated with some caution.* 

## Getting Started
TInsul can be installed using pip:

**pip install tinsul**

**Other required libraries**: numba, numpy, pandas

## Documentation
Current documentation can be found [here](https://tinsul.readthedocs.io/en/latest/).

## Examples
The main function within TInsul is called "tinsul_sim". It requires only one input (an array of temperatures), although a number of other optional inputs can also be supplied. "temps" is a 12x3 matrix, with the rows being months of the year (starting with January), and the columns being the low, average, and high temperatures per month, respectively. Alternatively, supplying "default" to the temps parameter will use the average monthly temperatures for Washington DC. "start_month" refers to the month at which to begin the simulation, with 1 being January, 2 being February, etc.

For example:

> temps = [[-2, 1, 6], [-1, 3, 8], [3, 7, 13], [8, 13, 19], [14, 18, 24], [19, 24, 29], [22, 28, 31], [21, 27, 30], [17, 22, 26], [10, 15, 20], [5, 10, 14], [0, 4, 8]]

> start_month = 1

> df = tinsul_sim(temps, start_month)

The output of the tinsul_sim function will be a DataFrame containing the condition indicators for paper insulation failure (CO, CO2, Furan, Furfural, Water Content). The size of the array can be used to infer the failure time, in weeks.

## Technical Basis
![Overview](/images/TransformerSimulationOverview.png)

First, the ambient temperature will be estimated based on the current month and the monthly lows, averages, and highs specified by the user. Per day, it is assumed that the transformer experiences 6 hours at the low temperature, 12 hours at the average temperature, and 6 hours at the high temperature.

Next, the core hot spot temperature, paper insulation damage. and oil contamination is calculated. The core hot spot temperature requires the ambient temperature and overload ratio as inputs. The paper insulation degradation module will use the core hot spot temperature to assess the damage to the paper insulation. The oil contamination module will take the current DP of the paper insulation and find the total accumulated CO and CO2.

Finally, the simulation must assess whether the paper insulation of the transformer has failed or not. Rather than set a “hard” threshold at 150 DP (as recommended by Emsley), the simulation was designed to give the paper insulation a probability to fail at values between 150 DP and 250 DP, resulting in a “soft” distribution of failures.

The transformer simulation contains three primary modules:

* Core hot spot
* Paper insulation degradation
* Oil contamination

The following sections will provide more detail about each module.

### Core Hot Spot
The degree to which paper insulation is degraded depends on the stress it receives from ambient heat. The "core hot spot" temperature of a transformer is the hottest temperature generated by the windings, and it represents the "weakest link" of the paper insulation. The simulation will only consider the probability that the part of the paper insulation closest to the core hot spot will tear, since it is at the point of highest stress.

The core hot spot temperature is not usually measured directly, but it can be inferred from the ambient temperature using empirical relationships. The equation used for this simulation is described in the paper "Loading Transformers By Temperature" [Montsinger]:

![Montsinger](/images/MontsingerEquation.png)

Where T is the total core hot spot temperature, T0 is the temperature rise over the ambient oil temperature, N is the copper to iron loss ratio, R is the ratio of the overload to the rated load, Tc is the temperature rise of the windings over the oil, and L is a known constant. All temperatures are in Celsius.

### Paper Insulation Degradation
The degree of polymerization (DP) of the paper insulation is one of the best indicators of its tensile strength. When first manufactured, the DP of paper insulation will typically be between 1000 and 1300. At paper insulation failure, its DP is reduced to about 20% of its starting value, and DP below 150 is almost guaranteed to fail immediately.

The general law of paper insulation degradation over time is given by [Emsley]:

![Emsley](/images/EmsleyEquation.png)

Where k is a constant that depends on the core hot spot temperature and physical characteristics of the transformer, t is the time in hours, DPf is the degree of polymerization of the paper after t hours, and DPi is the initial DP of the paper before t hours.

### Oil Contamination
Damage to the paper insulation will release dissolved gas byproducts (such as hydrogen, carbon monoxide, and carbon dioxide) in the insulating oil. These dissolved gases are condition indicators, correlated with the amount of degradation that the paper insulation has experienced.

For the oil contamination module, several papers were used to simulate the accumulation of carbon monoxide and carbon dioxide in a transformer based on paper insulation DP. Unfortunately, academic papers describing this relationship are limited, and power companies are very protective of their DGA data. Hence, measurements from a paper by [Pradhan and Ramu] were used to create a linear relationship between DP and carbon monoxide/dioxide.

While this relationship may not be entirely accurate, it is good enough for the purposes of this simulation. To account for the random variation in the paper's data, normally-distributed noise was added in this module based on the standard deviation of the empirical data.

This situation is also commonly modeled using chemical equations as a "physics-of-failure" model. Paper insulation is an organic compound, and its breakdown should result in hydrogen and carbon compounds, etc.

The equations developed from empirical data in Pradhan and Ramu's paper are:

![Pradhan](/images/PradhanEquation.png)

Note that the CO and CO2 values that this function returns are the total accumulation of these dissolved gases over time; it is not the current rate of accumulation.

## References
* Montsinger, V. M. "Loading Transformers By Temperature." AIEE Transactions 49 (1930): 776-79.

* Emsley, A. M., and G. C. Stevens. "Review of Chemical Indicators of Degradation of Cellulosic Electrical Paper Insulation in Oil-Filled Transformers." IEE Proc.-Sci. Meas. Technol.141.5 (1994).

* Pradhan, M. K., and T. S. Ramu. "On the Estimation of Elapsed Life of Oil-Immersed Power Transformers." IEEE Transactions on Power Delivery 20.3 (2005).

CHANGELOG.rst
============
 Change Log
============

Version 0.11
------------
- Initial release


tinsul/core.py
# -*- coding: utf-8 -*-
"""
    tinsul.core
    ~~~~~~~~~~
    A module to simulate condition-monitoring data for Transformer INSULation 
    prognostics models.

    :copyright: (c) 2017 by Eric Strong.
    :license: Refer to LICENSE.txt for more information.
"""

import numpy as np
import pandas as pd
from numba import jit
from numpy.random import randn


@jit
# Transformer INSULation SIMulator
def tinsul_sim(temps, start_month=1.0, dp_initial=1000, fail_loc=175.0,
               fail_scale=10.0, o=1.0, t0=35.0, tc=30.0, n=1.0, n0=0.8, nc=1.0,
               l=1.0, a=(3.7*10**7)):
    """
    tinsul_sim is a function to simulate condition-monitoring data for 
    Transformer INSULation prognostics models.
    
    :param temps: 12x3 array, containing monthly low, average, and high temps
            supplying "default" will use the temperatures for DC
    :param start_month: first week of January=1, second week=1.25, etc.
    :param dp_initial: starting degree of polymerization (DP) of the paper
    :param fail_loc: "location" parameter of the logistic failure distribution
    :param fail_scale: "scale" parameter of the logistic failure distribution
    :param o: overload_ratio, typically between 0.75 and 1.1 [Montsinger]
            if a list is supplied, simulation will iterate through the list
            as if it were a repeating pattern
    :param t0: temperature rise of oil over ambient [Montsinger]
    :param tc: temperature rise of the windings over the oil [Montsinger]
    :param n: ratio of copper to iron loss at rated load [Montsinger]
    :param n0: 0.8 for self-cooled transformers, 0.5 water-cooled [Montsinger]
    :param nc: 1.0 for vertical windings, 0.8 for horizontal [Montsinger]
    :param l: if hot spot is constant, l=1.0 [Montsinger]
    :param a: a constant based on the type of paper insulation [Emsley]
    :return: A DataFrame of condition indicators per week: co, co2, furan,
            furfural, and paper water content (in order)
    """
    if type(o) is float:
        o_list = [o]
    else:
        o_list = o
    if temps == "default":
        temps = _use_default_temps()
    # Stores the accumulated condition indicator values in a dict
    acc = {}
    # Variables to store some "state" values per iteration
    cur_month = start_month
    cur_dp = dp_initial
    cur_week = 0
    cur_o = 0
    # dp_failed determines the DP at which the paper insulation will
    # fail, drawn from a logistic distribution centered at 200
    fail_dp = np.random.logistic(loc=fail_loc, scale=fail_scale)
    while cur_week < 5000:  # transformers don't live longer than 5000 weeks
        # Simulate 6 hours at low, 12 hours at avg, and 6 hours at high temps
        # The first index is temp, the second index is hours at that temp
        ambient_low = [temps[int(cur_month)-1][0], 6]
        ambient_avg = [temps[int(cur_month)-1][1], 12]
        ambient_high = [temps[int(cur_month)-1][2], 6]
        # Update DP based on the heat stresses from the core hot spot
        for ambient, time in [ambient_low, ambient_avg, ambient_high]:
            chs = _core_hot_spot(ambient, o_list[cur_o], t0, tc, n, n0, nc, l)
            cur_dp = _calculate_dp(chs, time, cur_dp, a)
        # Calculate the condition indicators based on DP
        acc[cur_week] = _oil_contamination(cur_dp)
        # Check for transformer failure (less than 150 DP is instant failure)
        if (cur_dp <= fail_dp) | (cur_dp < 150):
            break
        # Add the value (Months/Weeks) to get proportion of month per week
        cur_month += 0.230769
        cur_week += 1
        cur_o += 1
        # Rollover to the next year, if necessary
        if cur_month >= 13.0:
            cur_month = 1
        if cur_o > (len(o_list) - 1):
            cur_o = 0
    # Convert the dict to a pandas DataFrame
    df = pd.DataFrame.from_dict(acc, orient='index')
    df.columns = ['CO', 'CO2', 'Furan', 'Furfural', 'Water Content']
    return df


# Uses the monthly temperatures from Washington, DC
def _use_default_temps():
    return [[-2, 1, 6],  # January
            [-1, 3, 8],  # February
            [3, 7, 13],  # March
            [8, 13, 19],  # April
            [14, 18, 24],  # May
            [19, 24, 29],  # June
            [22, 28, 31],  # July
            [21, 27, 30],  # August
            [17, 22, 26],  # September
            [10, 15, 20],  # October
            [5, 10, 14],  # November
            [0, 4, 8]]  # December

@jit
# Calculates the core hot spot temperature of the transformer
def _core_hot_spot(amb, o=1.0, t0=35.0, tc=30.0, n=1.0, n0=0.8, nc=1.0, l=1.0):
    # ---Refer to Montsinger's paper for more information---
    # amb is ambient temperature in Celsius
    # o is overload ratio in decimal form (e.g. 0.75 not 75%)
    # t0 is the temperature rise of oil over ambient
    # tc is the temperature rise of the windings over the oil
    # n = ratio of copper to iron loss at rated load (1 for simplification)
    # n0 = 0.8 for self-cooled transformers, 0.5 for water-cooled
    # nc = 1.0 for vertical windings, 0.8 for horizontal windings
    # l = hot spot is constant, kva varies with the ambient -> l=1.0
    t_chs = t0*((n*o**2+1)**n0) + tc*l*o**(2*nc) + amb
    return t_chs


@jit
def _calculate_dp(core_hot_spot, time, dp_initial, a=(3.7*10**7)):
    # ---Refer to Emsley's paper for more information---
    # core_hot_spot is from the _core_hot_spot function (in Celsius)
    # time is measured in hours
    # dp_initial is the initial degree of polymerization
    # a is a constant based on the type of paper insulation:
    #    -upgraded paper: 3.7*10**7
    #    -dry Kraft paper: 1.1*10**8
    #    -Kraft paper + 1% water: 3.5*10**8
    #    -Kraft paper + 2% water: 7.8*10**8
    #    -Kraft paper + 4% water: 3.5*10**9
    k = a * np.exp(-(117000 / (8.314 * (core_hot_spot + 273.15))))
    dp_final = 1 / ((k * 24 * 7 * time) + (1 / dp_initial))
    return dp_final


@jit
def _oil_contamination(dp):
    # dp is the degree of polymerization
    # ------------------------------------------------------------------------
    # This section contains estimates of dissolved gases and other features
    #  based on regression of empirical data in academic papers
    # ---Refer to Pradhan and Ramu's paper for more information---
    # CO and CO2 are the TOTAL accumulation of the gas, not the rate
    co = (-0.0028*dp + 6.28) + (0.13*randn())
    co2 = (-0.0026*dp + 8.08) + (0.66*randn())
    # ---Refer to "On the Estimation of Elapsed Life" for more information---
    furan = (-0.0038*dp + 7.93) + (0.13*randn())
    # ---Refer to "Study on the Aging Characteristics and Bond-Breaking
    # Process of Oil - Paper Insulation" for more information---
    furfural = (-0.0025*dp + 4.72) + (0.11*randn())
    # ---Refer to Emsley's paper for more information---
    water_content = (0.5 * np.log(1000 / dp)) / (np.log(2)) + (0.03*randn())
    return co, co2, furan, furfural, water_content
